clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -409,6 +409,9 @@ public void poll(long timeout) throws IOException {
         // Add to completedReceives after closing expired connections to avoid removing
         // channels with completed receives until all staged receives are completed.
         addToCompletedReceives();
+
+        // Clear all selected keys so that they are included in the ready count for the next select
+        this.nioSelector.selectedKeys().clear();
```

Comment:
Wouldn't it be better to do this right after the last `pollSelectionKeys`? We can even just do `readyKeys.clear()` then.

Reply:
`pollSelectionKeys` is called multiple times with different key sets. Not all of them need clearing. We only really need to clear `nioSelector.selectedKeys()`. I thought it was clearer this way.

Reply:
Misread the comment earlier, updated.